File: blogi_projekti_csharp/API/Services/BlogService.cs
using API.CustomExceptions;
using API.Data;
using API.Dtos;
using API.Interfaces;
using API.Models;
using Microsoft.EntityFrameworkCore;

namespace API.Services
{
    public class BlogService(DataContext _repository) : IBlogService
    {
        public async Task<IEnumerable<Blog>> GetAll()
        {
            var blogs = await _repository.Blogs.ToListAsync();
            return blogs;
        }

        public async Task<Blog> GetById(int id)
        {
            var blogWithOwner = await _repository
                .Blogs.Include(b => b.Owner)
                .Include(b => b.Tags)
                .FirstOrDefaultAsync(b => b.Id == id);
            if (blogWithOwner == null)
            {
                throw new NotFoundException("blog not found");
            }

            return blogWithOwner;
        }

        public async Task<Blog> Create(CreateBlogReq requestData, int loggedInUser)
        {
            // Haetaan jo oo. tagit, jotka löytyvät Tags-taulusta

            var existingTags = await _repository
                .Tags.Where(tag => requestData.Tags.Contains(tag.TagText))
                .ToListAsync();

            // katsotaan, onko joukossa uusia tageja, joita ei ole vielä tallennettu tietokantaan

            var existingTagTexts = existingTags.Select(tag => tag.TagText).ToList();
            var newTagTexts = requestData.Tags.Except(existingTagTexts).ToList();

            var newTagEntities = new List<Tag>();

            // lisätään uudet tagit listaan
            foreach (var text in newTagTexts)
            {
                newTagEntities.Add(new Tag { TagText = text });
            }
            // addataan kaikki uudet tagit kerralla
            await _repository.Tags.AddRangeAsync(newTagEntities);

            // yhdistetään uudet ja vanhat tägit yhteen listaan
            var allTags = existingTags.Union(newTagEntities).ToList();

            var blog = new Blog
            {
                Title = requestData.Title,
                Content = requestData.Content,
                AppUserId = loggedInUser,
                // lisätään tägit blogiin
                Tags = allTags,
            };

            // EfCoressa kaikki tietokantaan lisättävät rivit pitää ensin lisätä commitoitavien listaan käyttäen addia
            await _repository.Blogs.AddAsync(blog);
            // kun rivit on ensin lisätty listaan, muutokset voidaan tallentaa
            await _repository.SaveChangesAsync();
            return blog;
        }

        public async Task<Blog> Edit(int id, UpdateBlogReq requestData)
        {
            var blog = await GetById(id);

            blog.Title = requestData.Title;
            blog.Content = requestData.Content;

            await _repository.SaveChangesAsync();

            return blog;
        }

        public async Task Delete(int id)
        {
            var blog = await GetById(id);
            _repository.Blogs.Remove(blog);
            await _repository.SaveChangesAsync();
        }
    }
}

File: blogi_projekti_python/services/blog_service.py
import traceback
from typing import List

from sqlalchemy.orm import Query, Session

import dtos.blogs
import models
from custom_exceptions.not_found_exception import NotFoundException
from services.abc_blog_service import ABCBlogService


class BlogService(ABCBlogService):
    def __init__(self, _repository: Session):
        self._repository = _repository

    def get_all(self) -> List[models.Blogs]:
        query: Query = self._repository.query(models.Blogs)
        result = query.all()
        return result

    def get_by_id(self, _id: int) -> List[models.Blogs]:
        return self._repository.query(models.Blogs).filter_by(Id=_id).first()

    def create(
        self, req: dtos.blogs.CreateBlogReq, logged_in_user_id: int
    ) -> models.Blogs:

        try:
            existing_tags_qry: Query = self._repository.query(models.Tags).filter(
                models.Tags.TagText.in_(req.tags)
            )
            existing_tags: List[models.Tags] = existing_tags_qry.all()
            existing_tag_texts = [tag.TagText for tag in existing_tags]
            new_tag_texts = list(set(req.tags) - set(existing_tag_texts))

            new_tags: List[models.Tags] = []
            for text in new_tag_texts:
                _new_tag = models.Tags(TagText=text)

                new_tags.append(_new_tag)

            self._repository.add_all(new_tags)
            self._repository.commit()

            all_tags = existing_tags + new_tags
        except:
            traceback.print_exc
            raise Exception("Handling the tags on blog creation failed")

        try:
            category: models.Categorys
            existing_category_qry: Query = self._repository.query(
                models.Categorys
            ).filter(models.Categorys.name == req.category)
            if (existing_category := existing_category_qry.first()) is not None:
                print(
                    "did not create new category on blog creation",
                    "\nnimi: ",
                    existing_category.name,
                    "\nid: ",
                    existing_category.id,
                    "\nowner_id: ",
                    existing_category.owner_id,
                )
                category = existing_category
            else:
                new_category = models.Categorys(
                    name=req.category,
                    owner_id=logged_in_user_id,
                )
                self._repository.add(new_category)
                self._repository.commit()
                print(
                    "created new category on blog creation",
                    "\nnimi: ",
                    new_category.name,
                    "\nid: ",
                    new_category.id,
                    "\nowner_id: ",
                    new_category.owner_id,
                )
                category = new_category
        except:
            traceback.print_exc
            raise Exception("Handling the categorys on blog creation failed")

        try:
            blog = models.Blogs(
                Title=req.title,
                Content=req.content,
                CategoryId=category.id,
                AppUserId=logged_in_user_id,
                Tags_=all_tags,
            )
            self._repository.add(blog)
            self._repository.commit()
        except:
            traceback.print_exc
            raise Exception("Handling the blogs on blog creation failed")

        return blog

    def remove(self, _id: int):
        blog = self.get_by_id(_id)
        if blog is None:
            raise NotFoundException("blog not found")
        self._repository.delete(blog)
        self._repository.commit()
